pkg/gardener/shoot/hyperscaler/aws/zones.go

```diff
@@ -28,46 +33,115 @@ cidr: 10.250.0.0/16
     public: 10.250.160.0/20
     internal: 10.250.176.0/20
 */
-func generateAWSZones(workerCidr string, zoneNames []string) []v1alpha1.Zone {
+
+func generateAWSZones(workerCidr string, zoneNames []string) ([]v1alpha1.Zone, error) {
+	numZones := len(zoneNames)
+	if numZones < minNumberOfZones || numZones > maxNumberOfZones {
+		return nil, errors.New("Number of networking zones must be between 1 and 8")
+	}
+
 	var zones []v1alpha1.Zone
 
-	cidr, _ := netip.ParsePrefix(workerCidr)
-	workerPrefixLength := cidr.Bits() + workersBits
-	workerPrefix, _ := cidr.Addr().Prefix(workerPrefixLength)
+	cidr, err := netip.ParsePrefix(workerCidr)
+	if err != nil {
+		return zones, errors.Wrap(err, "failed to parse worker CIDR")
+	}
+
+	orgWorkerPrefixLength := cidr.Bits() + workersBits
+	workerPrefix, err := cidr.Addr().Prefix(orgWorkerPrefixLength)
+	if err != nil {
+		return zones, errors.Wrap(err, "failed to get worker prefix")
+	}
 
 	// delta - it is the difference between "public" and "internal" CIDRs, for example:
 	//    WorkerCidr:   "10.250.0.0/19",
 	//    PublicCidr:   "10.250.32.0/20",
 	//    InternalCidr: "10.250.48.0/20",
-	// 4 * delta  - difference between two worker (zone) CIDRs
+	// 4 * delta  - difference between two worker (zone) CIDRs 4096 hosts
+	// small_delta and smallPrefixLength are used for subnets created for last 5 (from 4th to 8th) zones
+	// small_delta  - difference between two subnetworks CIDRs for last 5 (from 4th to 8th) zones 1024 hosts
+	//    WorkerCidr:   "10.250.192.0/22",
+	//    PublicCidr:   "10.250.196.0/22",
+	//    InternalCidr: "10.250.200.0/22",
+
+	smallPrefixLength := orgWorkerPrefixLength + 3
+
 	delta := big.NewInt(1)
-	delta.Lsh(delta, uint(lastBitNumber-workerPrefixLength))
+	delta.Lsh(delta, uint(lastBitNumber-orgWorkerPrefixLength))
+	small_delta := big.NewInt(1)
+	small_delta.Rsh(delta, 2)
 
 	// base - it is an integer, which is based on IP bytes
 	base := new(big.Int).SetBytes(workerPrefix.Addr().AsSlice())
 
-	for _, name := range zoneNames {
-		zoneWorkerIP, _ := netip.AddrFromSlice(base.Bytes())
-		zoneWorkerCidr := netip.PrefixFrom(zoneWorkerIP, workerPrefixLength)
+	processed := make(map[string]bool)
+
+	for i, name := range zoneNames {
+		if _, ok := processed[name]; ok {
+			return nil, errors.Errorf("zone name %s is duplicated", name)
+		}
+		processed[name] = true
+
+		var workPrefixLength, publicPrefixLength, internalPrefixLength int
+		var deltaStep *big.Int
 
-		base.Add(base, delta)
-		base.Add(base, delta)
-		publicIP, _ := netip.AddrFromSlice(base.Bytes())
-		public := netip.PrefixFrom(publicIP, workerPrefixLength+1)
+		if i < 3 {
```

Comment:
Could you extract the magic number `3` to a const? E.g. with a name `zonesWithBiggerSubnets`.
You could also consider right above the constant name to link to https://github.com/kyma-project/infrastructure-manager/issues/887 that will provide some context why this behavior is different for the first 3 zones

Reply:
maybe something like `const kymaWorkerPoolAZs = 3 //first 3 zones are reserved for Kyma worker pool and bigger dimensioned than later added AZs`